### PR TITLE
Extend tests for bboxes datasets

### PR DESCRIPTION
**Pull request by sfmig**
## Description

**What is this PR**
- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**
To extend existing tests to bboxes datasets.

**What does this PR do?**
This PR covers the tests from the `logging` and `report` modules.

There are changes to the dataset validators tests too, because of refactoring a fixture involved in them.

## References
- Overflow from PR #229 
- Kinematic tests are extended in #265  

## How has this PR been tested?

Tests pass locally and in CI.

## Is this a breaking change?

No.

## Does this PR require an update to the documentation?

No.

## Checklist:

- [x] The code has been tested locally
- [x] Tests have been added to cover all new functionality
- [ n/a ] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -213,6 +214,29 @@ def sleap_file(request):
...
+
+@pytest.fixture
+def valid_bboxes_arrays_all_zeros():
+    """Return a dictionary of valid zero arrays (in terms of shape) for a
+    ValidBboxesDataset.
+    """
+    # define the shape of the arrays
+    n_frames, n_individuals, n_space = (10, 2, 2)
+
+    # build a valid array for position or shape will all zeros
```
Comment (niksirbi) [inline tests/conftest.py:228]:
```suggestion
# build a valid array for position or shape with all zeros
```

File: tests/test_unit/test_logging.py [hunk 2]
```diff
@@ -45,27 +46,42 @@ def test_log_warning(caplog):
...
-@pytest.mark.parametrize("input_data", ["dataset", "dataarray"])
-def test_log_to_attrs(input_data, valid_poses_dataset):
+@pytest.mark.parametrize(
+    "input_data",
+    [
+        "valid_poses_dataset",
+        "valid_bboxes_dataset",
+    ],
+)
+@pytest.mark.parametrize(
```
Comment (niksirbi) [inline tests/test_unit/test_logging.py:56]: Neat trick

File: tests/test_unit/test_logging.py [hunk 3]
```diff
@@ -45,27 +46,42 @@ def test_log_warning(caplog):
...
+    # check the log in the dataset is as expected
+    assert isinstance(output_data, expected_selector_type)
     assert "log" in output_data.attrs
     assert output_data.attrs["log"][0]["operation"] == "fake_func"
-    assert (
-        output_data.attrs["log"][0]["arg_1"] == "test1"
-        and output_data.attrs["log"][0]["kwarg"] == "test2"
-    )
+    assert output_data.attrs["log"][0]["arg_1"] == "test1"
+    assert output_data.attrs["log"][0]["kwarg"] == "test2"
```
Comment (niksirbi) [inline tests/test_unit/test_logging.py:86]: the exact same line is repeated twice?

File: tests/test_unit/test_reports.py [hunk 4]
```diff
@@ -4,28 +4,136 @@
...
-        ),  # Missing "individuals" and "keypoints" dims
+        "valid_poses_dataset",
+        "valid_bboxes_dataset",
+        "valid_poses_dataset_with_nan",
+        "valid_bboxes_dataset_with_nan",
     ],
 )
-def test_report_nan_values(
-    capsys, valid_poses_dataset_with_nan, data_selection
+@pytest.mark.parametrize(
```
Comment (niksirbi) [inline tests/test_unit/test_reports.py:15]: I personally prefer to have the args in the functions signature in the same order as in the parametrisation, and things like capsys and request at the end.

File: tests/test_unit/test_reports.py [hunk 5]
```diff
@@ -4,28 +4,136 @@
...
+    "data_selection, list_expected_individuals_idcs",
```
Comment (niksirbi) [inline tests/test_unit/test_reports.py:16]: shoud `idcs` be `ids` or `idxs`? Feels like there is a stray `c`.

File: tests/test_unit/test_reports.py [hunk 6]
```diff
@@ -4,28 +4,136 @@
...
+    and only the relevant individuals are included in the report.
+    """
+    # apply selection to extract relevant position data
+    input_dataset = request.getfixturevalue(valid_dataset)
+    output_data_array = data_selection(input_dataset)
+
+    # produce report
+    report_str = report_nan_values(output_data_array)
+
+    # check report of nan values includes dataset name
```
Comment (niksirbi) [inline tests/test_unit/test_reports.py:44]: that's the data array name, right? not the dataset name

Comment (niksirbi) [reply tests/test_unit/test_reports.py:44]: both in the comment and in the assertion error message.

File: tests/test_unit/test_reports.py [hunk 7]
```diff
@@ -4,28 +4,136 @@
...
+)
+@pytest.mark.parametrize(
+    "data_selection, list_expected_keypoints, list_expected_individuals",
+    [
+        (
+            lambda ds: ds.position,
+            ["key1", "key2"],
+            ["ind1", "ind2"],
+        ),
+        # Report nans for all keypoints and individuals
```
Comment (niksirbi) [inline tests/test_unit/test_reports.py:80]: I was confused by the comment coming after the corresponding code block, I'd go for putting the comment above the set of parameters it refers to

File: tests/test_unit/test_reports.py [hunk 8]
```diff
@@ -4,28 +4,136 @@
...
+            ["ind1", "ind2"],
+        ),
+        # Report nans for keypoint "key1" only, and all individuals
+        # if only one keypoint: not explicitly reported in output
+        (
+            lambda ds: ds.position.sel(individuals="ind1", keypoints="key1"),
+            [],
+            ["ind1"],
+        ),
+        # Report nans for individual "ind1" and keypoint "key1" only
```
Comment (niksirbi) [inline tests/test_unit/test_reports.py:93]: there is an extra "only" here

File: tests/test_unit/test_reports.py [hunk 9]
```diff
@@ -4,28 +4,136 @@
...
+    valid_dataset,
+    data_selection,
+    list_expected_keypoints,
+    list_expected_individuals,
+    request,
 ):
-    """Test that the nan-value reporting function handles data
-    with missing ``individuals`` and/or ``keypoint`` dims, and
-    that the dataset name is included in the report.
+    """Test that the nan-value reporting function handles position data
```
Comment (niksirbi) [inline tests/test_unit/test_reports.py:105]: This docstring appears to be copy-pasted from the above test. References to "individuals" should be changed to "keypoints"

File: tests/test_unit/test_reports.py [hunk 10]
```diff
@@ -4,28 +4,136 @@
...
-    assert data.name in report_nan_values(
-        data
+    # apply selection to extract relevant position data
+    input_dataset = request.getfixturevalue(valid_dataset)
+    output_data_array = data_selection(input_dataset)
+
+    # produce report
+    report_str = report_nan_values(output_data_array)
+
+    # check report of nan values includes dataset name
```
Comment (niksirbi) [inline tests/test_unit/test_reports.py:116]: data array name (if changed above also)

File: tests/test_unit/test_validators/test_datasets_validators.py [hunk 11]
```diff
@@ -274,15 +259,21 @@ def test_bboxes_dataset_validator_with_invalid_shape_array(
         (
             [1, 2, 3],
             pytest.raises(ValueError),
-            "Expected 'individual_names' to have length 2, but got 3.",
-        ),  # length doesn't match position_array.shape[1]
+            "Expected 'individual_names' to have length 2, "
+            f"but got {len([1, 2, 3])}.",
+        ),
+        # length doesn't match position_array.shape[1]
```
Comment (niksirbi) [inline tests/test_unit/test_validators/test_datasets_validators.py:265]: same comment as before, I find it more natural to have the comment before the code

File: tests/test_unit/test_validators/test_datasets_validators.py [hunk 12]
```diff
@@ -313,13 +304,15 @@ def test_bboxes_dataset_validator_individual_names(
         (
             np.ones((10, 3, 2)),
             pytest.raises(ValueError),
-            "Expected 'confidence_array' to have shape (10, 2), "
-            "but got (10, 3, 2).",
-        ),  # will not match position_array shape
+            f"Expected 'confidence_array' to have shape (10, 2), "
+            f"but got {np.ones((10, 3, 2)).shape}.",
+        ),
+        # will not match shape of position_array in
```
Comment (niksirbi) [inline tests/test_unit/test_validators/test_datasets_validators.py:310]: same here

**Review by niksirbi — APPROVED** [review]
Well done on extending the tests! I particularly like how you also test subsets of data with the selectors.

I just had some cosmetics/pedantic comments.

File: tests/test_unit/test_logging.py (hunk 2 above)
Comment (niksirbi) [reply tests/test_unit/test_logging.py:56]: I mean with passing the lambdas as selectors

File: tests/test_unit/test_logging.py (hunk 3 above)
Comment (niksirbi) [reply tests/test_unit/test_logging.py:86]: line 86 and 87

Comment (sfmig) [reply tests/test_unit/test_logging.py:86]: yes, thanks for spotting it! 

Sometimes vscode does this with the last line in a file, I _think_ when I ctrl+s to force save a file. I still have no idea why, but there is at least [one report ](with the last line in the file)of a similar bug

File: tests/test_unit/test_reports.py (hunk 5 above)
Comment (sfmig) [reply tests/test_unit/test_reports.py:16]: I actually meant `idcs` as indices, plural of index.

I meant to distinguish it from IDs.

Will spell it out for clarity

File: tests/test_unit/test_reports.py (hunk 4 above)
Comment (sfmig) [reply tests/test_unit/test_reports.py:15]: Same, I think `capsys` is actually a typo cause it's not used

File: tests/test_unit/test_reports.py (hunk 6 above)
Comment (sfmig) [reply tests/test_unit/test_reports.py:44]: yes, that was inherited from the previous test - fixing now

File: tests/test_unit/test_reports.py (hunk 5 above)
Comment (niksirbi) [reply tests/test_unit/test_reports.py:16]: Ah, I typically use `idxs` for the same thing (and `idx` for singular). I had just never seen `idcs`

**Pull request merged**